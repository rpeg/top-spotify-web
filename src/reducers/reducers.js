import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
