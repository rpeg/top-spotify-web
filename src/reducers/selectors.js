export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
