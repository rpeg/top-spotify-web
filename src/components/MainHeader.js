import React from 'react';
import { useSelector } from 'react-redux';
import { Container, Row, Col } from 'react-bootstrap';
import Moment from 'moment';

import Profile from './Profile';
import { selectUser } from '../reducers/selectors';

const MainHeader = () => {
  const user = useSelector(selectUser);

  return (
    user && user.id ? (
      <Container style={{ paddingTop: '2em' }}>
        <Row>
          <Col xs={6} style={{ padding: '0', display: 'block' }}>
            <span
              style={{
                display: 'inline-flex',
                alignItems: 'center',
                float: 'left',
              }}
            >
              <img
                style={{ height: '50px', width: 'auto' }}
                src={`${process.env.PUBLIC_URL}/Spotify_Icon_RGB_Green.png`}
                alt="Spotify Logo"
              />
              <div style={{ width: 'auto', margin: '0 0 0 0.5em' }}>
                <p className="text-m" style={{ textAlign: 'left', margin: '0' }}><b>My Top Music</b></p>
                <p className="text-s" style={{ textAlign: 'left', margin: '0' }}>
                  {Moment().format('MMM. Do, YYYY')}
                </p>
              </div>
            </span>
          </Col>
          <Col xs={6} style={{ padding: '0', display: 'block' }}>
            <Profile user={user} />
          </Col>
        </Row>
      </Container>
    ) : null
  );
};

export default MainHeader;
