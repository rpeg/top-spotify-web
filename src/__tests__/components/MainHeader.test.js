import React from 'react';
import { shallow } from 'enzyme';

import { Container } from 'react-bootstrap';
import MainHeader from '../../components/MainHeader';
import * as selectors from '../../reducers/selectors';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = ({ user = {} }) => {
  jest.spyOn(selectors, 'selectUser').mockReturnValue(user);
};

describe('MainHeader', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    setup({});
    const wrapper = shallow(<MainHeader />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly without user', () => {
    setup({});
    const wrapper = shallow(<MainHeader />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Container).exists()).toBe(false);
  });

  it('renders correctly with user', () => {
    const user = {
      id: 'abc',
    };

    setup({ user });

    const wrapper = shallow(<MainHeader />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Container).exists()).toBe(true);
  });
});
