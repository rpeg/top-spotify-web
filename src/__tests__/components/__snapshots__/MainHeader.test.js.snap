// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainHeader renders correctly with user 1`] = `
<Container
  fluid={false}
  style={
    Object {
      "paddingTop": "2em",
    }
  }
>
  <ForwardRef
    noGutters={false}
  >
    <Col
      style={
        Object {
          "display": "inline-flex",
          "padding": "0",
        }
      }
      xs={6}
    >
      <span
        style={
          Object {
            "alignItems": "center",
            "display": "inline-flex",
          }
        }
      >
        <img
          alt="Spotify Logo"
          src="/Spotify_Icon_RGB_Green.png"
          style={
            Object {
              "height": "50px",
              "width": "auto",
            }
          }
        />
        <div
          style={
            Object {
              "margin": "0 0 0 0.5em",
              "width": "auto",
            }
          }
        >
          <h4
            style={
              Object {
                "margin": "0",
                "textAlign": "left",
              }
            }
          >
            My Top Music
          </h4>
          <div>
            <p
              style={
                Object {
                  "margin": "0",
                  "textAlign": "left",
                }
              }
            >
              Dec. 4th, 2019
            </p>
          </div>
        </div>
      </span>
    </Col>
    <Col
      style={
        Object {
          "display": "inline-flex",
          "padding": "0",
        }
      }
      xs={6}
    >
      <Profile
        user={
          Object {
            "id": "abc",
          }
        }
      />
    </Col>
  </ForwardRef>
</Container>
`;

exports[`MainHeader renders correctly without user 1`] = `""`;
